File: packages/sdk/src/core/queue/batch-sender.ts
import type { HttpClient } from '../client/http-client';
import type { BatchItem, BatchRequest } from '../types';
import { VALIDATION } from '../types';
import { logger } from '../utils/logger';
import type { OfflineQueue } from './offline-queue';

const FLUSH_TIMEOUT_MS = 5000;

export class BatchSender {
  private readonly httpClient: HttpClient;
  private readonly offlineQueue: OfflineQueue;
  private isFlushing = false;

  constructor(httpClient: HttpClient, offlineQueue: OfflineQueue) {
    this.httpClient = httpClient;
    this.offlineQueue = offlineQueue;
  }

  async flush(): Promise<void> {
    if (this.isFlushing) {
      logger.warn('Flush already in progress. Skipping duplicate flush.');
      return;
    }

    this.isFlushing = true;

    try {
      await this.flushWithTimeout();
    } finally {
      this.isFlushing = false;
    }
  }

  private async flushWithTimeout(): Promise<void> {
    const timeoutPromise = new Promise<void>((_, reject) => {
      setTimeout(() => reject(new Error('Flush timeout')), FLUSH_TIMEOUT_MS);
    });

    const flushPromise = this.performFlush();

    try {
      await Promise.race([flushPromise, timeoutPromise]);
    } catch (error) {
      logger.error(
        'Flush failed or timed out. Dropping remaining items.',
        error
      );
    }
  }

  private async performFlush(): Promise<void> {
    const items = await this.offlineQueue.dequeueAll();

    if (items.length === 0) {
      return;
    }

    const deduplicatedItems = this.deduplicateByTimestamp(items);
    const batches = this.splitIntoBatches(deduplicatedItems);

    for (const batch of batches) {
      try {
        await this.sendBatch(batch);
      } catch (error) {
        logger.error(
          'Batch send error. Dropping batch (fire & forget).',
          error
        );
      }
    }
  }

  private async sendBatch(items: BatchItem[]): Promise<void> {
    const itemsWithoutRetryCount = items.map(({ retryCount, ...item }) => item);
    const request: BatchRequest = { items: itemsWithoutRetryCount };
    const result = await this.httpClient.sendBatch(request);

    if (!result.success) {
      logger.warn('Batch request failed. Dropping batch (fire & forget).', {
        error: result.error,
      });
      return;
    }

    const response = result.data;
    if (response.failed && response.failed > 0) {
      logger.warn(
        `Batch partially failed: ${response.failed}/${(response.processed ?? 0) + response.failed} items dropped.`
      );
    }
  }

  private deduplicateByTimestamp(items: BatchItem[]): BatchItem[] {
    const seenEvents = new Map<string, number>();
    const dedupedItems: BatchItem[] = [];
    const DEDUP_WINDOW_MS = 50;

    for (const item of items) {
      if (item.type === 'event') {
        const key = this.createEventKey(item.payload.name, item.payload.params);

        let timestamp: number;
        try {
          timestamp = new Date(item.payload.timestamp).getTime();
          if (Number.isNaN(timestamp)) {
            logger.warn('Invalid timestamp format. Including event anyway.');
            dedupedItems.push(item);
            continue;
          }
        } catch {
          logger.warn('Failed to parse timestamp. Including event anyway.');
          dedupedItems.push(item);
          continue;
        }

        const lastTime = seenEvents.get(key);
        if (lastTime !== undefined && timestamp - lastTime < DEDUP_WINDOW_MS) {
          logger.warn('Duplicate event in batch detected. Dropping event.', {
            name: item.payload.name,
          });
          continue;
        }

        seenEvents.set(key, timestamp);
        dedupedItems.push(item);
      } else {
        dedupedItems.push(item);
      }
    }

    const droppedCount = items.length - dedupedItems.length;
    if (droppedCount > 0) {
      logger.info(
        `Dropped ${droppedCount} duplicate event(s) from batch based on timestamp.`
      );
    }

    return dedupedItems;
  }

  private createEventKey(
    name: string,
    params?: Record<string, string | number | boolean | null>
  ): string {
    if (!params) {
      return name;
    }
    try {
      const sortedParams = Object.keys(params)
        .sort()
        .reduce(
          (acc, key) => {
            acc[key] = params[key];
            return acc;
          },
          {} as Record<string, string | number | boolean | null>
        );
      return `${name}:${JSON.stringify(sortedParams)}`;
    } catch {
      return name;
    }
  }

  private splitIntoBatches(items: BatchItem[]): BatchItem[][] {
    const batches: BatchItem[][] = [];
    const maxSize = VALIDATION.BATCH.MAX_SIZE;

    for (let i = 0; i < items.length; i += maxSize) {
      batches.push(items.slice(i, i + maxSize));
    }

    return batches;
  }
}

File: packages/swift/Sources/PhaseAnalytics/Core/Queue/BatchSender.swift
import Foundation

internal actor BatchSender {
    private let httpClient: HTTPClient
    private let offlineQueue: OfflineQueue
    private var isFlushing = false

    private static let maxRetryCount = 5

    init(httpClient: HTTPClient, offlineQueue: OfflineQueue) {
        self.httpClient = httpClient
        self.offlineQueue = offlineQueue
    }

    func flush() async {
        guard !isFlushing else {
            logger.warn("Flush already in progress. Skipping duplicate flush.")
            return
        }

        isFlushing = true
        defer { isFlushing = false }

        let items = await offlineQueue.dequeueAll()
        guard !items.isEmpty else {
            return
        }

        let deduplicatedItems = deduplicateByTimestamp(items)
        let batches = splitIntoBatches(deduplicatedItems)

        for batch in batches {
            let result = await sendBatch(batch)

            if case .failure(let error) = result {
                logger.error("Batch request failed. Will retry.", error)
                await requeue(batch)
            }
        }
    }

    private func sendBatch(_ items: [BatchItem]) async -> Result<BatchResponse, PhaseError> {
        let request = BatchRequest(items: items)
        let result = await httpClient.sendBatch(request)

        if case .success(let response) = result {
            if response.failed > 0 {
                let total = (response.processed ?? 0) + response.failed
                logger.error("Batch partially failed: \(response.failed)/\(total) items")

                var failedItems: [BatchItem] = []

                for error in response.errors {
                    guard let clientOrder = error.clientOrder else {
                        logger.error("Error missing clientOrder. Cannot re-enqueue.")
                        continue
                    }

                    if let failedItem = items.first(where: { $0.clientOrder == clientOrder }) {
                        failedItems.append(failedItem)
                    }
                }

                if !failedItems.isEmpty {
                    await requeue(failedItems)
                }
            }
        }

        return result
    }

    private func requeue(_ items: [BatchItem]) async {
        await withTaskGroup(of: Void.self) { group in
            for item in items {
                group.addTask {
                    let retryCount = (item.retryCount ?? 0) + 1

                    guard retryCount <= Self.maxRetryCount else {
                        logger.error("Max retries exceeded. Dropping item.")
                        return
                    }

                    let newItem: BatchItem
                    switch item {
                    case .device(let payload, let order, _):
                        newItem = .device(payload: payload, clientOrder: order, retryCount: retryCount)
                    case .session(let payload, let order, _):
                        newItem = .session(payload: payload, clientOrder: order, retryCount: retryCount)
                    case .event(let payload, let order, _):
                        newItem = .event(payload: payload, clientOrder: order, retryCount: retryCount)
                    case .ping(let payload, let order, _):
                        newItem = .ping(payload: payload, clientOrder: order, retryCount: retryCount)
                    }

                    await self.offlineQueue.enqueue(newItem)
                }
            }
        }
    }

    private func deduplicateByTimestamp(_ items: [BatchItem]) -> [BatchItem] {
        var seenEvents: [String: Date] = [:]
        var dedupedItems: [BatchItem] = []
        let dedupWindow: TimeInterval = 0.05

        let iso8601Formatter = ISO8601DateFormatter()
        iso8601Formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

        for item in items {
            if case .event(let payload, _, _) = item {
                let key = createEventKey(name: payload.name, params: payload.params)

                guard let timestamp = iso8601Formatter.date(from: payload.timestamp) else {
                    logger.warn("Invalid timestamp format. Including event anyway.")
                    dedupedItems.append(item)
                    continue
                }

                if let lastTime = seenEvents[key], timestamp.timeIntervalSince(lastTime) < dedupWindow {
                    logger.warn("Duplicate event in batch detected. Dropping event: \(payload.name)")
                    continue
                }

                seenEvents[key] = timestamp
                dedupedItems.append(item)
            } else {
                dedupedItems.append(item)
            }
        }

        let droppedCount = items.count - dedupedItems.count
        if droppedCount > 0 {
            logger.info("Dropped \(droppedCount) duplicate event(s) from batch based on timestamp.")
        }

        return dedupedItems
    }

    private func createEventKey(name: String, params: [String: AnyCodable]?) -> String {
        guard let params = params else {
            return name
        }

        let sortedParams = params.sorted { $0.key < $1.key }
        if let jsonData = try? JSONSerialization.data(
            withJSONObject: sortedParams.reduce(into: [String: Any]()) { dict, pair in
                dict[pair.key] = pair.value.value
            },
            options: [.sortedKeys]
        ),
            let jsonString = String(data: jsonData, encoding: .utf8)
        {
            return "\(name):\(jsonString)"
        }

        return name
    }

    private func splitIntoBatches(_ items: [BatchItem]) -> [[BatchItem]] {
        let maxSize = ValidationConstants.Batch.maxSize
        var batches: [[BatchItem]] = []

        for i in stride(from: 0, to: items.count, by: maxSize) {
            let endIndex = min(i + maxSize, items.count)
            batches.append(Array(items[i..<endIndex]))
        }

        return batches
    }
}
